File: src/roar_net_api/api/operations/random_move.py
# SPDX-FileCopyrightText: © 2025 Authors of the roar-net-api-py project <https://github.com/roar-net/roar-net-api-py/blob/main/AUTHORS>
#
# SPDX-License-Identifier: Apache-2.0

from typing import Protocol, TypeVar, Optional

Solution = TypeVar("Solution", contravariant=True)
Move = TypeVar("Move", covariant=True)


class SupportsRandomMove(Protocol[Solution, Move]):
    def random_move(self, solution: Solution) -> Optional[Move]: ...

File: julia/src/operations/random_move.jl

# SPDX-FileCopyrightText: © 2025 Authors of the roar-net-api-py project <https://github.com/roar-net/roar-net-api-py/blob/main/AUTHORS>
#
# SPDX-License-Identifier: Apache-2.0
"""
    random_move(neighbourhood::Neighbourhood, solution::Solution) -> move::Union{Nothing, Move}

Return a single random `Move` from `neighbourhood` for `solution`,
or `nothing` if no move exists.
"""
function random_move(::Neighbourhood, ::Solution)::Union{Nothing, Move} end
